reinforced_lib/rlib.py

```diff
@@ -367,40 +378,88 @@ def fit(self, agent_id: int = 0) -> None:
-            The identifier of agent instance.
         path : str
-            Path to the output file.
+            Path to the checkpoint file. If ``.pkl.lz4`` extension not detected, it will be appended automaticly.
         """
 
-        if not path:
-            raise ValueError('No path is specified.')
-
-        # TODO implement objects saving and agree on a format
-
-        raise NotImplementedError()
-
-    def load_agent_state(self, path) -> int:
+        if path is None:
+            path = os.path.join(os.path.expanduser("~"), f"checkpoint{self._lz4_ext}")
+        elif path[-8:] != self._lz4_ext:
+            path = path + self._lz4_ext
+
+        experiment_state = {
+            "agent_type": self._agent_type,
+            "agent_params": self._agent_params,
+            "agents": {
+                agent_id: {
+                    "state": self._agents_states[agent_id],
+                    "key": self._agents_keys[0],
+                } for agent_id in range(len(self._agents_states))
+            },
+            "ext_type": self._ext_type,
+            "ext_params": self._ext_params,
+            "loggers_type": self._loggers_type,
+            "loggers_sources": self._loggers_sources,
+            "loggers_params": self._loggers_params
+        }
+
+        with lz4.frame.open(path, 'wb') as f:
+            f.write(pickle.dumps(experiment_state))
+    
+    @staticmethod
+    def load(
+        path: str,
+        agent_params: Dict[str, Any] = None,
+        ext_params: Dict[str, Any] = None,
+    ) -> RLib:
         """
-        Loads agent instance from file in X format.
+        Loads the state of the experiment from a file in lz4 format.
 
         Parameters
         ----------
         path : str
-            Path to the input file.
-
-        Returns
-        -------
-        id : int
-            The identifier of loaded instance.
+            Path to the checkpoint file.
+        agent_params : Dict[str, Any], optional
+            Dictionary of altered agents parameters with their new values, by default None
+        ext_params : Dict[str, Any], optional
+            Dictionary of altered extension parameters with their new values, by default None
         """
+        
+        with lz4.frame.open(path, 'rb') as f:
+            experiment_state = pickle.loads(f.read())
+        
+        rlib = RLib()
+        
+        rlib._agents_states = []
+        rlib._agents_keys = []
+
+        if ext_params:
+            rlib.set_ext(experiment_state["ext_type"], ext_params)
+        else:
+            rlib.set_ext(experiment_state["ext_type"], experiment_state["ext_params"])
+        
+        if agent_params:
+            rlib.set_agent(experiment_state["agent_type"], agent_params)
+        else:
+            rlib.set_agent(experiment_state["agent_type"], experiment_state["agent_params"])
 
-        # TODO implement loading objects from file
+        if experiment_state["loggers_type"]:
```

Comment:
Na ten moment nie robiliśmy zmiany typów i parametrów loggerów dla uproszczenia i to jest okey, ale gdyby użytkownik chciał zmienić logger, to nie powinniśmy mu tego zabraniać. Myślę, że można dodać na ten moment prostą flagę `restore_loggers`, którą będzie można ustawić żeby nie przywracała loggerów i wtedy użytkownik może sobie sam ustawić przez wywołanie `set_loggers`.

Reply:
ok, jest to jakieś rozwiązanie